setup.py

```diff
@@ -280,6 +281,26 @@ def run(self):
             print(f"Copying {file} to {dst_file}")
             self.copy_file(file, dst_file)
 
+        # copy vllm_flash_attn/layers dir
+        src_layers = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
+                                  "layers")
+        dst_layers = os.path.join("vllm", "vllm_flash_attn", "layers")
+        if os.path.exists(src_layers):
+            print(f"Copying directory {src_layers} to {dst_layers}")
+            if os.path.exists(dst_layers):
+                shutil.rmtree(dst_layers)
+            shutil.copytree(src_layers, dst_layers)
+
+        # copy vllm_flash_attn/ops dir
+        src_ops = os.path.join(self.build_lib, "vllm", "vllm_flash_attn",
+                               "ops")
+        dst_ops = os.path.join("vllm", "vllm_flash_attn", "ops")
+        if os.path.exists(src_ops):
+            print(f"Copying directory {src_ops} to {dst_ops}")
+            if os.path.exists(dst_ops):
+                shutil.rmtree(dst_ops)
+            shutil.copytree(src_ops, dst_ops)
```

Comment:
You should use `self.copy_tree` instead

```suggestion
				for folder in ("layers", "ops"):
            src = os.path.join(self.build_lib, "vllm", "vllm_flash_attn", folder)
            out = os.path.join("vllm", "vllm_flash_attn", folder)
            self.copy_tree(src, out)
```

Reply:
You can use this patch here

```patch
diff --git a/setup.py b/setup.py
index 671ec7739..f5b3f5165 100755
--- a/setup.py
+++ b/setup.py
@@ -283,6 +283,13 @@ class cmake_build_ext(build_ext):
             dst_file = os.path.join("vllm/vllm_flash_attn", basename)
             print(f"Copying {file} to {dst_file}")
             self.copy_file(file, dst_file)
+        # We will need to copy these folders to use its ops for qwen-vl
+        for folder in ("layers", "ops"):
+            src = os.path.join(self.build_lib, 'vllm', 'vllm_flash_attn',
+                               folder)
+            out = os.path.join('vllm', 'vllm_flash_attn', folder)
+            print(f"Copying {folder} from vllm/vllm_flash_attn")
+            self.copy_tree(src, out)
 
 
 class repackage_wheel(build_ext):
@@ -402,6 +409,32 @@ class repackage_wheel(build_ext):
 
                 package_data[package_name].append(file_name)
 
+            # Extract and include the layers and ops folders for qwen-vl
+            folders_to_copy = {"layers", "ops"}
+            for folder in folders_to_copy:
+                folder_path = f"vllm/vllm_flash_attn/{folder}"
+                folder_files = [
+                    f for f in wheel.filelist
+                    if f.filename.startswith(folder_path)
+                ]
+
+                if folder_files:
+                    print(f"Include {folder} folder from vllm/vllm_flash_attn")
+                    for file in folder_files:
+                        wheel.extract(file)
+
+                        # Add the file to package_data if it's not a Python file
+                        rel_path = file.filename.split("/")
+                        # vllm/vllm_flash_attn/folder/file
+                        if len(rel_path) >= 4:
+                            package_name = "vllm.vllm_flash_attn." + folder
+                            file_name = rel_path[-1]
+
+                            if not file_name.endswith(".py"):
+                                if package_name not in package_data:
+                                    package_data[package_name] = []
+                                package_data[package_name].append(file_name)
+
 
 def _is_hpu() -> bool:
     # if VLLM_TARGET_DEVICE env var was set explicitly, skip HPU autodetection
```

Reply:
Thank you